File: DonRyu/2022-10-11-House-Robber.js
var rob = function(nums) {
    if (nums.length === 0) return 0;
    if (nums.length === 1) return nums[0]

    // Keep track of the max money we can make with x amount of houses available
    // dp[0] = max amount if we only have the first house to rob
    // dp[1] = max amount if we only have the first 2 houses to rob
    let dp = [nums[0], Math.max(nums[0], nums[1])];

    for (let i = 2; i < nums.length; i++) {
        // Compare current max with the previous max
        // Check if the money from the current house + max of 2 houses away is greater than the current max
        dp[i] = Math.max(dp[i-2] + nums[i], dp[i-1]);
    }
    return dp[nums.length - 1];
};

File: BrianLee/Week5/2022-10-11-House-Robber.java
class Solution {
    public int rob(int[] nums) {
        int[] mem = new int[nums.length];
        for(int i = 0; i < nums.length; i++)
            mem[i] = -1;

        return fun(nums.length-1, nums, mem);
    }

    private int fun(int n, int[] nums, int[] mem) {
        if(n < 0) return 0;
        if(n == 0) return nums[0];
        if(n == 1) return Math.max(nums[0], nums[1]);

        if(mem[n] != -1) return mem[n];
        mem[n-2] = fun(n-2, nums, mem);
        mem[n-1] = fun(n-1, nums, mem);

        return Math.max(mem[n-2] + nums[n], mem[n-1]);
    }
}